shopinvader_payment/components/payment_transaction_event_listerner.py

```diff
@@ -0,0 +1,38 @@
+# Copyright 2019 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo.addons.base_rest.controllers.main import _PseudoCollection
+from odoo.addons.component.core import Component
+from odoo.addons.shopinvader import shopinvader_response
+
+
+class SaleOrderPaymentTransactionEventListener(Component):
+    _name = "sale.order.payment.transaction.event.listener"
+    _inherit = "base.event.listener"
+    _apply_on = ["sale.order"]
+
+    def _confirm_and_invalidate_session(self, sale_order):
+        shopinvader_backend = sale_order.shopinvader_backend_id
+        if not shopinvader_backend:
+            return
+        sale_order.action_confirm_cart()
+        response = shopinvader_response.get()
+        response.set_session("cart_id", 0)
+        response.set_store_cache("cart", {})
+        # TODO we should not have to return the last_sale information into the
+        # response, only the id...
+        # This code is an awful hack... We should never have to call
+        # a service implementation from here. That should be the responsibility
+        # of the client to request the cart info to store into ist cache once
+        # once the cache is reset
+        collection = _PseudoCollection("shopinvader.backend", self.env)
+        work = self.work.work_on(collection=collection)
+        setattr(work, "shopinvader_backend", shopinvader_backend)
+        res = work.component(usage="cart")._to_json(sale_order)
+        response.set_store_cache("last_sale", res.get("data", {}))
+        # end of awful code ....
+
+    def on_payment_transaction_pending(self, sale_order, transaction):
+        self._confirm_and_invalidate_session(sale_order)
```

Comment:
Not sure about this point.
Indeed in previous version (as we only use transaction for CB/paypal case) the sale order was confirm only when the transaction was done : https://github.com/shopinvader/odoo-shopinvader-payment/blob/10.0/shopinvader_payment/services/abstract_payment.py#L86

What we need it to confirm the cart when
- for SIPS, Stripe, paypal... the state is done
- for check, bank transfer ... the state is pending (before we didn't had this case)

Maybe we can have an option on the payment mode? or maybe on shopinvader_payment object?

Reply:
@sebastienbeau I know, it's a little bit fragile and works because the pending state is not used by SIPS and stripe. Maybe an option on payment mode or??? Can you create an issue for this topic to not bloc the PR and collect ideas from others?

Reply:
Sadly Stripe can have pending state on transaction : https://github.com/shopinvader/odoo-shopinvader-payment/pull/7/files#diff-d431953f29bacf6a3149cc925ca1013fR17

Reply:
Note: before it's was not configurable but depend on provider (stripe, sips => always done)

Reply:
The pending state is now only used if the provider is "transfer"
This case is implemented into shopinvader_payment_manual since it's a specific case link to this kind of payment.